flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessor.java

```diff
@@ -157,4 +186,66 @@ private void fullCheckAndSetAvailable() {
             }
         }
     }
+
+    /** Visible for testing only. Do not use out side StreamMultipleInputProcessor. */
+    @VisibleForTesting
+    public static class MultipleInputAvailabilityHelper {
+        private final CompletableFuture<?>[] cachedAvailableFutures;
+        private final Consumer[] onCompletion;
+        private CompletableFuture<?> availableFuture;
+
+        public CompletableFuture<?> getAvailableFuture() {
+            return availableFuture;
+        }
+
+        public static MultipleInputAvailabilityHelper newInstance(int inputSize) {
+            MultipleInputAvailabilityHelper obj = new MultipleInputAvailabilityHelper(inputSize);
+            return obj;
+        }
+
+        private MultipleInputAvailabilityHelper(int inputSize) {
+            this.cachedAvailableFutures = new CompletableFuture[inputSize];
+            this.onCompletion = new Consumer[inputSize];
+        }
+
+        @VisibleForTesting
+        public void init() {
+            for (int i = 0; i < cachedAvailableFutures.length; i++) {
+                final int inputIdx = i;
+                onCompletion[i] = (Void) -> notifyCompletion(inputIdx);
+            }
+        }
+
+        public boolean isInvalid() {
+            return availableFuture == null || availableFuture.isDone();
+        }
+
+        public void resetToUnavailable() {
+            availableFuture = new CompletableFuture<>();
+        }
+
+        private void notifyCompletion(int idx) {
+            if (availableFuture != null && !availableFuture.isDone()) {
+                availableFuture.complete(null);
+            }
+            cachedAvailableFutures[idx] = AVAILABLE;
+        }
```

Comment:
My guess would be this code might be the reason behind the deadlocks. `notifyCompletion()` is executed from different threads (whoever is completing the `StreamOneInputProcessor` availability future) while it's accessing non thread safe fields here.

I haven't yet thought how it should be fixed, but we should avoid adding extra synchronisation overhead in this code.

Reply:
I think this code has still a potential for a deadlock, as the thread calling this `notifyCompletion()` method can see an outdated value  of the `availableFuture` field.

Reply:
`notifyCompletion` may competing with `checkReusableAndReset`. We could not prevent that from happening. 
But, this competing may be not harmful.
In your fix, the `AtomicReference` can not prevent that neither.  The `isDone()` check is still seperate with the reset.
Both ours fix is working. I have test them with 300 test runs. Did not see anykind of hang-up.

Reply:
| Benchmark |                                  Mode |  Cnt |      Score |     Error |   Units |
| -------- | ------------| -------- | -------- | ------ | ------- |
|MultipleInputBenchmark.multiInputMapSink |  thrpt |  300 |  16874.869 | ± 327.760 |  ops/ms |

Reply:
> In your fix, the AtomicReference can not prevent that either.

It doesn't need to. Take a closer look on the order of execution in my version methods `maybeReset()` and `registerFuture()` and think about what will happen if one of the input's availability future will complete concurrently. 
1. If input becomes available before we call `maybeReset()` it's future will be hooked up to the new combined future.
2. If input becomes available just after `anyAvailable.get().isDone()` check, but before `anyAvailable.set(new CompletableFuture<>())`, the combined/returned future will be available anyway. We will clean up everything in the next `getAvailableFuture()` call.
3. If input becomes available after `maybeReset()`, it will just complete the combined/returned regardless if we register it or not.
4. `registerFuture()` calls will take care of setting up all of the inputs' futures.

So I don't see any race condition in my version. Yours works quite similar after all, but your version's lack of `AtomicReference` creates an opportunity, for example in point 3., that if input becomes available, it will attempt to complete wrong, obsolete, old, already completed future.

Reply:
> So I don't see any race condition in my version. Yours works quite similar after all, but your version's lack of AtomicReference creates an opportunity, for example in point 3., that if input becomes available, it will attempt to complete wrong, obsolete, old, already completed future.

The race condition exists in both of our fix. The `AtomicReference` without `CAS` operations is no difference with plain field reference. We both relies on redundant completion check to make sure that we are not missing any completion notification.

I'm quite sure, replace the `AtomicReference` with plain field will make no difference.
But this it not saying that your fix is not working. But only some opinion to the `AtomicReference` usage.

Reply:
> The AtomicReference without CAS operations is no difference with plain field reference.

I'm afraid you are wrong in this regard. All atomic classes are wrappers around `volatile` field + indeed some extra atomic CAS operations and such [they provide happens before relationship:](https://www.tutorialspoint.com/java_concurrency/concurrency_atomicreference.htm)

> That is, a set has a happens-before relationship with any subsequent get on the same variable. The atomic compareAndSet method also has these memory consistency features.

Reply:
In this scenario, we may not benefits a lot from the `volatile` field. Since, it only preventing dirty memory cache between CPU cores. 

> So I don't see any race condition in my version. Yours works quite similar after all, but your version's lack of AtomicReference creates an opportunity, for example in point 3., that if input becomes available, it will attempt to complete wrong, obsolete, old, already completed future.

Even with `volatile`, the future completion can still happens inbetween the `maybeReset()` call and the completion callback will see the obsolete, old, already completed `anyAvailable` future, and try to complete it.
It's no difference. The AtomicReference is not prevent this. AtomicReference's `set` and `get` are plain method on plain object, without the `CAS` operation. The only thing difference here is the `volatile` field. But, since we can not preventing the race condition here. Future completion still has the chance to see old future object. So, `volatile` is the least thing to concern here.
But, anyway, there is no harm to make the `availableFuture` `volatile`.

Reply:
> Even with volatile, the future completion can still happens inbetween the maybeReset() call (let's say, after the isDone() check, and before the set operation) and the completion callback will see the obsolete, old, already completed anyAvailable future, and try to complete it.
> It's no difference. The AtomicReference is not preventing this. AtomicReference's set and get are plain method on plain object. 

As I was trying to explain above in the point 2.

> Take a closer look on the order of execution in my version methods maybeReset() and registerFuture()

If input becomes available just after `anyAvailable.get().isDone()` check, but before `anyAvailable.set(new CompletableFuture<>())`, the combined/returned future will be available anyway. Such input would return `true` for  `needsToBeRegistered()` since its future would be completed, `registerFuture()` would be executed, which would immediately complete the combined future.

Reply:
I don't say your fix is not working. I just say it still works even without `AtomicReference`. So, did my fix here.

Reply:
Maybee check on this. I draw a graph to illustrate how this fix is correct.
> I made a graph to explain how this fix works. Even with the race condition between future's `notifyCompletion` and the `checkReusableAndReset`.
> ![flink-issue-timeline drawio](https://user-images.githubusercontent.com/2689362/151551134-a30c100e-26a9-4061-9467-6c9b03657bcc.png)

Reply:
The only divergence here is the on the use of `AtomicReference`. 
My option is that `AtomicReference` is not crucial in this scenario. Yours seems to be that the `AtomicReference` is the missing piece of my fix which preventing getting a correct results.
I think both of our fix are functioning. I will make some last modification to the concern you have about the `volatile` feature.